### ISSUE TITLE
"top k" for Metric Meta

### ISSUE THREAD

**Issue by Dav1dde** [getsentry/relay#2787]
Our current approach for metric meta:

Every relay instance remembers the first N (currently 5) code locations per MRI per project per day, every additional code location is discarded. If the relay sees a new code location it passes it upwards to the next relay layer up until it hits processing relays.
Processing relays have the same in-memory deduplication logic but instead of passing the metadata upstream they persist it in Redis.

We store one key per project, mri, timestamp and type of metric metadata (currently only locations). The locations key is an (unbounded) set of seen locations. There is no maximum size etc. check on these sets, while technically they could grow to infinity, practically this does not happen due to the de-duplication we do in relay.

This approach allows us to heavily de-duplicate in Relay already with no coordination between relay instances. Also on average we should get a nice distribution of metric locations following to the distribution of releases in the wild (relevant for client SDKs, e.g. mobile).

tldr; we store the first five locations per day per metric.

---

There are some potential problems with this approach, we cannot rank code locations except by a first seen timestamp. Meaning we can't specifically show the user the metric location we have seen most in the UI. Another issue is that if there are more than N (5) code locations for one metric we might be discarding an important location because we already saw 5 other locations first, though this should be a very rare and unlikely occasion.

To mitigate this the idea is to instead track the Top-N code locations via Redis Top-K and then show the top 5 results to the user instead.

Depending on future usecases of metric meta, Top-K can be a more general solution also covering these future usecases.

Considerations:
- We cannot as effectively discard metadata in relays anymore, at the very least all locations will need to hit processing relays
- A naive implementation would no longer aggregate or filter in memory but only use Redis, meaning every emitted code location will hit our Redis, while this is currently not a problem since the metadata is still very very low volume we have not seen how much data client side SDKs will send us. I expect this a considerable amount more than with the current Python SDK and Sentry pilot phase. This might not scale enough.
- Due to the last point, we will need some smart in memory aggregation. E.g. have a Top-K in memory and sync to Redis occasionally via [TOPK.INCRBY](https://redis.io/commands/topk.incrby/). This may lead to a more complex implementation.
- We need to investigate if Top-K will actually give us any real benefit over the very trivial implementation we are running now on the user side. 
- At some point we will switch to a permanent storage (from Redis) where we will have to re-think this again.

**Comment by Dav1dde** [comment]
Blocked on https://github.com/getsentry/sentry/issues/60684 for now. Then we will decide wether we move forward with this.

**Comment by Dav1dde** [comment]
Let's close this, maybe relevant when we have a new storage.
